crates/outbound-mqtt/src/lib.rs

```diff
@@ -97,16 +97,18 @@ impl v2::HostConnection for OutboundMqtt {
                 .await
                 .map_err(other_error)?;
 
-            // Poll EventLoop once to send the message to MQTT broker or capture/throw error
-            // We may revisit this later to manage long running connections and their issues in the connection pool.
-            eventloop
-                .poll()
-                .await
-                .map_err(|err: rumqttc::ConnectionError| {
-                    v2::Error::ConnectionFailed(err.to_string())
-                })?;
-
-            Ok(())
+            // Poll event loop until outgoing publish event is iterated over to send the message to MQTT broker or capture/throw error.
+            // We may revisit this later to manage long running connections, high throughput use cases and their issues in the connection pool.
+            loop {
+                let event = eventloop
+                    .poll()
+                    .await
+                    .map_err(|err| v2::Error::ConnectionFailed(err.to_string()))?;
+
+                if let rumqttc::Event::Outgoing(rumqttc::Outgoing::Publish(_)) = event {
+                    return Ok(());
+                }
```

Comment:
I think we will need this part to have a bit more logic depending on the QoS of the published message. We may need to also check for `PubAck` if Qos level is 1. If QoS level is 2, then we need to wait for `PubComp`.

Reply:
Would something like this work here?

```
use rumqttc::{QoS, Event, Outgoing, Incoming};

loop {
    let event = eventloop
        .poll()
        .await
        .map_err(|err| v2::Error::ConnectionFailed(err.to_string()))?;

    match (qos, event) {
        (QoS::AtMostOnce, Event::Outgoing(Outgoing::Publish(_))) |
        (QoS::AtLeastOnce, Event::Incoming(Incoming::PubAck(_))) |
        (QoS::ExactlyOnce, Event::Outgoing(Outgoing::PubComp(_))) => break,

        (_, _) => continue,
    }
}
```

Reply:
...will have a look at it this weekend, test, update and create an issue to track parallel requests issue (para 1 of my [comment](https://github.com/fermyon/spin/pull/2362#discussion_r1526538543)) . Thanks.

Reply:
Code updated, issue #2407 created to track and fix forward. Thanks